package was.httpserver;

import java.io.PrintWriter;

import static java.nio.charset.StandardCharsets.UTF_8;

public class HttpResponse {

    private final PrintWriter writer;
    private int statusCode = 200;
    private final StringBuilder bodyBuilder = new StringBuilder();
    private String contentType = "text/html; charset=UTF-8";

    public HttpResponse(PrintWriter writer) {
        this.writer = writer;
    }

    public void setStatusCode(int statusCode) {
        this.statusCode = statusCode;
    }

    public void setContentType(String contentType) {
        this.contentType = contentType;
    }

    public void writeBody(String body) {
        bodyBuilder.append(body);
    }

    public void flush() {
        int contentLength = bodyBuilder.toString().getBytes(UTF_8).length;
        writer.println("HTTP/1.1 " + statusCode + " " + getResonPhrase(statusCode));
        writer.println("Content-Type: " + contentType);
        writer.println("Content-Length: " + contentLength);
        writer.println();
        writer.println(bodyBuilder.toString());
        writer.flush();
    }

    private String getResonPhrase(int statusCode) {
        switch (statusCode) {
            case 200:
                return "HTTP/1.1 " + statusCode + " " + "OK";
            case 404:
                return "HTTP/1.1 " + statusCode + " " + "Not Found";
            case 500:
                return "HTTP/1.1 " + statusCode + " " + "Internal Server Error";
            default: return "Unknown Status";
        }
    }
}
